hal/src/nRF52840/servo_hal.cpp

```diff
@@ -1,57 +1,51 @@
-  License as published by the Free Software Foundation, either
-  version 3 of the License, or (at your option) any later version.
-
-  This library is distributed in the hope that it will be useful,
-  but WITHOUT ANY WARRANTY; without even the implied warranty of
-  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-  Lesser General Public License for more details.
-
-  You should have received a copy of the GNU Lesser General Public
-  License along with this library; if not, see <http://www.gnu.org/licenses/>.
- ******************************************************************************
+/*
+ * Copyright (c) 2018 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
  */
 
-/* Includes ------------------------------------------------------------------*/
 #include "servo_hal.h"
+#include "pwm_hal.h"
+#include "pinmap_impl.h"
 
-/* Private macro -------------------------------------------------------------*/
+#define SERVO_PWM_RESOLUTION        15
+#define SERVO_PWM_MAX_VALUE         ((1 << SERVO_PWM_RESOLUTION) - 1)
 
-/* Private variables ---------------------------------------------------------*/
-
-/* Extern variables ----------------------------------------------------------*/
-
-/* Private function prototypes -----------------------------------------------*/
-
-void HAL_Servo_Attach(uint16_t pin)
-{
+void HAL_Servo_Attach(uint16_t pin) {
+    // Set PWM frequency to 50Hz
+    HAL_PWM_Write_With_Frequency(pin, 0, SERVO_TIM_PWM_FREQ);
 }
 
-void HAL_Servo_Detach(uint16_t pin)
-{
+void HAL_Servo_Detach(uint16_t pin) {
+    HAL_PWM_Reset_Pin(pin);
 }
 
-void HAL_Servo_Write_Pulse_Width(uint16_t pin, uint16_t pulseWidth)
-{
+void HAL_Servo_Write_Pulse_Width(uint16_t pin, uint16_t pulseWidth) {
+    HAL_PWM_Set_Resolution(pin, SERVO_PWM_RESOLUTION);
+    NRF5x_Pin_Info *pin_map = HAL_Pin_Map();
+    pin_map[pin].user_data = pulseWidth;
+
+    uint32_t period_us = 1000000 / SERVO_TIM_PWM_FREQ;
+    uint32_t pwm_duty_value = pulseWidth * SERVO_PWM_MAX_VALUE / period_us;
+    HAL_PWM_Write_With_Frequency_Ext(pin, pwm_duty_value, SERVO_TIM_PWM_FREQ);
 }
 
-uint16_t HAL_Servo_Read_Pulse_Width(uint16_t pin)
-{
-    return 0;
+uint16_t HAL_Servo_Read_Pulse_Width(uint16_t pin) {
+    NRF5x_Pin_Info* pin_map = HAL_Pin_Map();
+    return pin_map[pin].user_data;
```

Comment:
The conversion from analog value(0~255 or 0~0xFFFFFFFF) to microseconds will bring errors

Reply:
But this would be the actual pulse width though, wouldn't it? We can use higher resolution if needed as well.

Reply:
Yes, the servo pulse unit is microseconds(1000,000), the PWM resolution is only 15bit (0x1FFFF=131,071), if we want to git rid of errors, we need to change PWM resolution to a higher value. Although the resolution is software implement, the hardware truly supports up to 15bit resolution. If we change `MAX_RESOLUTION_BITS` to a higher value, the users will get inaccurate PWM wave which is not the same as they expected.

Reply:
I'm fine with leaving this as-is, but at the same time maybe from the user perspective having an actual pulse width (including the errors) is more beneficial instead of simply returning the value that was passed to `HAL_Servo_Write_Pulse_Width()`.

Reply:
Totally, it's a fake implementation. I found some example using `writeMicroseconds(value) == readMicroseconds(value)` to check if the operation is successful, in this case the condition will fail.

Reply:
Let's leave this as-is then.